Removed node dependencies, the project now uses axios (featuring both callbacks and promises)

SudoPlz opened the pull request.

SudoPlz committed:
- Fixed a typo.

SudoPlz committed:
The SDK now uses ES6 promises instead of callbacks.
- Changed both the basic and oAuth implementations.
- Changed the examples to use promises instead of callbacks.

SudoPlz committed:
Now using bluebird promises, plus we use both callbacks and promises.
- Replaced request-promise-native with request-promise.
- Added back the callback syntax, we now always return bluebird promises, but when there's a callback we also call that too.
- Removed some catch blocks to make the example code easier to read.
- Now the promises return the full response, instead of just the body.

SudoPlz committed:
- Fixed the examples, we were consuming the wrong value on the request returned object.

SudoPlz committed:
- Now we only throw an error when a callback function does not exist.

SudoPlz committed:
Added catch blocks to our basic example.

SudoPlz committed:
Replaced node crypto with crypto-js.

SudoPlz committed:
Replaced request lib with a more lightweight one (axios), mostly to make our sdk independent from node and it's dependencies.

cesutherland commented on src/AcuityScheduling.js:
```diff
@@ -2,60 +2,62 @@
   // Set configurable options:
   if (options.auth)     config.auth     = options.auth;
-  if (options.body)     config.body     = options.body;
+  if (options.data)     config.data     = options.data;
```
We'll want to keep the "body" parameter for backwards compatibility.

SudoPlz replied:
Ah yes, I'm on it!

cesutherland commented on src/AcuityScheduling.js:
```diff
@@ -2,60 +2,62 @@
-  if (options.body)     config.body     = options.body;
+  if (options.data)     config.data     = options.data;
   if (options.method)   config.method   = options.method;
-  if (options.qs)       config.qs       = options.qs;
```
We should keep this for backwards compatibility, too.  Looks fine to map it to the `params` option for axios?

SudoPlz replied:
I'll try that today and see if it works with the `params` option. Good catch.

cesutherland commented on src/AcuityScheduling.js:
```diff
@@ -2,60 +2,62 @@
+  if (options.data)     config.data     = options.data;
   if (options.method)   config.method   = options.method;
-  if (options.qs)       config.qs       = options.qs;
+  if (options.paramsSerializer) config.paramsSerializer = options.paramsSerializer;
```
Does this serialize `params` without a default?  If not we should provide a default.

SudoPlz replied:
It does serialize `params` by  default but I found out it serialises them to JSON. I now serialise it using `querystring`.

cesutherland commented on src/AcuityScheduling.js:
```diff
@@ -2,60 +2,62 @@
-    cb(err, response, body);
+  return axios(config)
+  .then(function(res){
+    if (cb) {cb(null, res, res.data);}
```
:+1:

cesutherland commented on src/AcuityScheduling.js:
```diff
@@ -2,60 +2,62 @@
+  })
+  .catch(function(err){
+    if (cb) {
+      cb(err.response.data);
```
This should match the old behavior, but I'll want to double check it.  At a glance, it looks like request/response for axios are different than request/request-promise :-/

SudoPlz replied:
It is, but the `err.response.data` object is exactly the `json` the server replies back to us. Do we need to return something other than that too?
By the way yesterday I ran all the examples and they seemed to be working fine both with promises and with callbacks.

cesutherland replied:
Got it!  So previously fatal things like network errors were set in err.  In that case, I think `err.response.data` won't exist.  Could you fix that?

If we had a valid error response from the server, err was not set and we returned the response and response.data like a successful API call.  Let's keep that same behavior.

The example page will look like this:
![image](https://cloud.githubusercontent.com/assets/348406/20083391/e87d3f94-a510-11e6-8ca7-2417c8211d69.png)

SudoPlz replied:
Got it! It should be gone now after https://github.com/SudoPlz/acuity-js/commit/7fa2696343b0e464f888cb4afe045d838c48db20

cesutherland commented on src/index.js:
```diff
@@ -24,9 +24,7 @@ var acuity = {
-    var hasher = crypto.createHmac('sha256', secret);
-    hasher.update(body);
-    var hash = hasher.digest('base64');
+    var hash = CryptoJS.HmacSHA256(body, secret).toString(CryptoJS.enc.Base64);
```
Looks good but I'll double check that this is correct.

SudoPlz replied:
I did check that, but go ahead and double check that would be good!!

cesutherland replied:
:+1: looks good!

cesutherland commented on examples/basic/index.js:
```diff
@@ -14,34 +14,34 @@ app.get('/', function (req, res) {
   var acuity = Acuity.basic(config);
   var response = res;
 
-  acuity.request('/me', function (err, res, me) {
```
Let's leave `examples/basic/index.js` using callbacks.

SudoPlz replied:
Ok I'll do that!

SudoPlz committed:
Better backwards compatibility. Reverted basic example to callbacks.

SudoPlz commented:
Ok, just changed the params that seemed to break stuff and reverted to the old basic example.

By the way, the review feature from github is A-MAZING.

SudoPlz committed:
If the user provides us with a qs param, it gets serialized as a querystring, instead of as a json.

cesutherland commented on examples/webhooks/index.js:
```diff
@@ -25,7 +25,7 @@ app.get('/', function (req, res) {
 app.post('/webhook', verifyMiddleware, function (req, res) {
   // The message is authentic:
-  console.log("The message is authentic:\n" + JSON.stringify(req.body, null, '  '));
+  console.log("The message is authentic:\n" + JSON.stringify(req.data, null, '  '));
```
This is a mistake — `req.body` is populated in express by the body parser.  We log it to the server console here :-)

SudoPlz replied:
Woops, you're right. Fixed https://github.com/SudoPlz/acuity-js/commit/7fa2696343b0e464f888cb4afe045d838c48db20

cesutherland commented on examples/custom-sidebar/index.js:
```diff
@@ -26,7 +26,7 @@ app.get('/', function (req, res) {
   setTimeout(function () {
     res.render('sidebar.html', {
-      details: JSON.stringify(req.body, null, '  ')
+      details: JSON.stringify(req.data, null, '  ')
```
This is also from the body-parser middleware in express, and should be `req.body`.

SudoPlz replied:
Ah yes, https://github.com/SudoPlz/acuity-js/commit/7fa2696343b0e464f888cb4afe045d838c48db20.

cesutherland commented on examples/create-appointment/index.js:
```diff
@@ -93,19 +94,19 @@ app.post('/', function (request, response) {
+      appointment: JSON.stringify(appointmentR.data, null, '  ')
     });
-  });
+  }).catch(function(err){return console.error("Error appointments: "+err.error+" with message: "+err.message);})
```
Somehow I'm hitting this with the log: `Error appointments: bad_request with message: Expected a JSON object.`

I haven't been able to get to the bottom of it yet, but this is another example it'd be nice to display errors in the browser for (instead of an endless spinner).  Don't worry much about this one though — I can fix it on the branch.

SudoPlz committed:
Server returned error responses no longer throw error objects, instead they get returned as plain objects.

SudoPlz committed:
Fixed an error where we returned the wrong value (data) when we were supposed to return the actual body of the express response.

cesutherland commented:
@SudoPlz merged!

SudoPlz commented:
Fantastic!!! :) 👍
